ab_decrypt.py

```diff
@@ -128,50 +124,97 @@ def masterKeyJavaConversion(k):
-    print( hexlify(toBytes2) )
-  ck2 = PBKDF2( toBytes2, header['mkSumSalt'], Nck, header['round'] ) 
+  if VERBOSITY>1:
+    dprint('PBKDF2 secret value for password verification is: ', end='')
+    dprint( hexlify(toBytes2) )
+  ck2 = PBKDF2( toBytes2, header['mkSumSalt'], Nck, header['round'] )
   if ck2!=ck:
-    print( 'computed ck:', hexlify(ck2), 'is different than embedded ck:', hexlify(ck) )
+    dprint( 'computed ck:', hexlify(ck2), 'is different than embedded ck:', hexlify(ck) )
   else:
-    print('password verification is OK')  
-  
-  encrypted = f.read()
-  if options.verbose:
-    print('decrypting',len(encrypted),'bytes... ', end='')
-  # decryption using master key and iv  
-  decryptedData = AES.new(mk, AES.MODE_CBC, iv).decrypt( encrypted )
-  if options.verbose:
-    print('OK')
-
-elif header['encryption']=='none':
-  print('no encryption') 
-  decryptedData = f.read()
-else:
-  print('unknown encryption')
-  exit()
-  
-if options.verbose:
-  print('decompression... ', end='')
-# decompression (zlib stream)  
-decompressed = decompress(decryptedData)
-if options.verbose:
-  print('done (',len(decompressed),'bytes )')
-print('writing backup as .tar ... ', end='')
-out = open(options.output,'wb')
-out.write(decompressed)
-print('OK. Filename is \''+options.output+'\'')  
-out.close()  
-  
-f.close()
\ No newline at end of file
+    dprint('password verification is OK')
+  # decryption using master key and iv
+  return AES.new(mk, AES.MODE_CBC, iv)
+
+def chunkReader(f, chunkSize=CHUNK_SIZE):
+  data = f.read(chunkSize)
+  while data:
+    yield data
+    data = f.read(chunkSize)
+
+def decrypt(encryptedIter, aesobj, chunkSize=CHUNK_SIZE):
+  for encrypted in encryptedIter:
+    yield aesobj.decrypt(encrypted)
+
+def decompress(compressedDataIter, blockSize=CHUNK_SIZE):
+  decompressobj = zlib.decompressobj()
+  for compressedData in compressedDataIter:
+    yield decompressobj.decompress(compressedData)
+  yield decompressobj.flush()
+  if not decompressobj.eof:
+    raise RuntimeError("incomplete or truncated zlib stream")
+
+def ab2tar(f, fout, password=None):
+  if f.readline()[:-1]!=b'ANDROID BACKUP':
+    dprint('not ANDROID BACKUP')
+    return False
+
+  #parse header
+  header = readHeader(f)
+  if VERBOSITY>1:
+    dprint(header)
+
+  if header['encryption']==b'AES-256':
+    if password is None:
+      password = inputtty("Enter Password: ")
+    password = password.encode('utf-8')
+    compressedIter = decrypt(chunkReader(f), getAESDecrypter(header, password))
+  elif header['encryption']==b'none':
+    dprint('no encryption')
+    compressedIter = chunkReader(f)
+  else:
+    dprint('unknown encryption')
+    return False
+
+  # decompression (zlib stream)
+  dprint('writing backup as .tar ... ', end='', flush=True)
+  for decData in decompress(compressedIter):
+    fout.write(decData)
+  dprint(f'OK. Filename is \'{fout.name}\', {fout.tell()} bytes written.')
+  out.close()
```

Comment:
You forgot to change `out` to `fout` here.

Reply:
After fixing the issue above, I ran diff to compare the output with that of master and they were identical.

Reply:
Good catch, not sure how I missed that. Thanks!